src/projection.jl

```diff
@@ -31,114 +36,317 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
+        end
+        return reshape(dy, project.axes)
+    end
 end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
-# Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+# Zero-dimensional arrays -- these have a habit of going missing:
+function (project::ProjectTo{AbstractArray})(dx::Number) # ... so we restore from numbers
+    project.axes isa Tuple{} || sum(length, project.axes) == 1 || throw(_projection_mismatch(project.axes, size(dx)))
+    fill(project.element(dx))
 end
-(project::ProjectTo{<:Array{T}})(dx::Array) where {T<:Number} = project.element.(dx)
-function (project::ProjectTo{<:Array{T}})(dx::AbstractZero) where {T<:Number}
-    return zeros(T, project.size)
+
+# Arrays of arrays -- store projector per element:
+ProjectTo(xs::AbstractArray{<:AbstractArray}) = ProjectTo{AbstractArray{AbstractArray}}(; elements=map(ProjectTo, xs), axes = axes(xs))
+function (project::ProjectTo{AbstractArray{AbstractArray}})(dx::AbstractArray)
+    dy = if axes(dx) == project.axes
+        dx
+    else
+        for d in 1:max(ndims(dx), length(project.axes))
+            size(dx, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(project.axes, size(dx)))
+        end
+        reshape(dx, project.axes)
+    end
+    # This always re-constructs the outer array, it's not super-lightweight
+    return map((f,x) -> f(x), project.elements, dy)
 end
-function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
-    return project(dx.parent)
+
+# Arrays of other things -- since we've said we support arrays, but may not support their elements,
+# we handle the container as above but store trivial element projector:
+ProjectTo(xs::AbstractArray) = ProjectTo{AbstractArray}(; element=ProjectTo(), axes=axes(xs))
+
+# Ref -- likewise aim at containers of supported things, but treat unsupported trivially.
+ProjectTo(x::Ref{<:Number}) = ProjectTo{Ref}(; x = ProjectTo(getindex(x)))
+ProjectTo(x::Ref{<:AbstractArray}) = ProjectTo{Ref}(; x = ProjectTo(getindex(x)))
+ProjectTo(x::Ref) = ProjectTo{Ref}(; x = ProjectTo())
+(project::ProjectTo{Ref})(dx::Ref) = Ref(project.x(dx[]))
+# And like zero-dim arrays, allow restoration from a number:
+(project::ProjectTo{Ref})(dx::Number) = Ref(project.x(dx))
+
+function _projection_mismatch(axes_x::Tuple, size_dx::Tuple)
+    size_x = map(length, axes_x)
+    DimensionMismatch("variable with size(x) == $size_x cannot have a gradient with size(dx) == $size_dx")
+end
+
+#####
+##### `LinearAlgebra`
+#####
+
+# Row vectors
+function ProjectTo(x::LinearAlgebra.AdjointAbsVec{T}) where {T<:Number}
+    sub = ProjectTo(parent(x))
+    ProjectTo{Adjoint}(; parent=sub)
+end
+# Note that while [1 2; 3 4]' isa Adjoint, we use ProjectTo{Adjoint} only to encode AdjointAbsVec.
```

Comment:
Why not name this `ProjectTo{AdjointAbsVec}` and avoid all possible confusion?

Reply:
Because `ProjectTo{Adjoint{T, <:AbstractVector} where T}` is ugly, and breaks the rule about using the head of the tree as the type.

I think it's ok, every `ProjectTo{T}` is really it's own little world under flag `T`, once you've decided you'll employ `T`, and what backing it goes with, then you must write what methods it wants. These don't fall out of the type, it's just a label.

Reply:
> Because `ProjectTo{Adjoint{T, <:AbstractVector} where T}` is ugly, and breaks the rule about using the head of the tree as the type.

`AdjointAbsVec` is a fake head of the type tree though.
kinda this does make me worry that this is a sign that just the head of the type tree may not always be enough?
In this case it is fine because we are not using the other.
But if we stick to it always we lose triangular dispatch, and are restricted to diagonal distpatch?
Getting out of that was a major feature of some past julia version.

But like you say it is mostly just a label for a little world